src/lib.rs

```diff
@@ -71,6 +71,59 @@ const OFFS_STACK_SIZE_ENTRIES: u64 = 16;
 pub struct SMRec {
     id: u64,            // Stackmap ID.
     offset: u32,        // Stackmap offset from start of containing func.
+    num_locs: u16,
+    locs: Vec<SMLoc>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct SMLoc {
+    kind: LocKind,
+    size: u16,
+    dwarf_reg: u16,
+    offset: LocOffset,
+}
+
+/// Unfortunately, due to a discrepancy between the llvm stackmap documentation
+/// [0] and the implementation of their own stackmap parser [1], we need this
+/// enum to interpret the integer type differently depending on its `LocKind`.
+/// The offset is interpreted as a u32 *iff* the `LocKind` is a Constant. In all
+/// other cases, this is an i32. There are examples [2] in the LLVM test suite
+/// where the offset value contains an integer which won't fit inside an i32.
+/// For now, we interpret this in the same way that llvm-readobj, and its test
+/// suite expects.
+///
+/// [0] https://llvm.org/docs/StackMaps.html#id10
+/// [1] https://github.com/llvm/llvm-project/blob/57b38a8593bd7d63b9db09676087365d8d3d0d8a/llvm/include/llvm/Object/StackMapParser.h#L123
+/// [2] https://github.com/llvm/llvm-project/blob/master/llvm/test/CodeGen/X86/stackmap-large-location-size.ll
+///
+// #XXX: Update this when we get clarification on what the correct behaviour is
+// from the LLVM devs.
+#[derive(Debug, Eq, PartialEq)]
+pub enum LocOffset {
+    I32(i32),
+    U32(u32)
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum LocKind {
+    Register,
+    Direct,
+    Indirect,
+    Constant,
+    ConstIndex
+}
+
+impl LocKind {
+    fn from_hex(val: u8) -> SMParserResult<LocKind>{
+        match val {
+            0x1 => Ok(LocKind::Register),
```

Comment:
Is the `0x` prefix necessary? Is it for consistency with how LLVM specify things?

Reply:
Not necessary no, just for consistency with how they specify it in the docs

Reply:
OK.